patches/tModLoader/Terraria/ModLoader/ChestLoot.cs

```diff
@@ -0,0 +1,217 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Terraria.ID;
+
+namespace Terraria.ModLoader
+{
+	public readonly struct ChestLoot {
+		public struct Group {
+			public readonly Item item;
+			private Func<bool> condition;
+
+			public Group(Item item, Func<bool> condition = null) {
+				this.item = item;
+				this.condition = condition;
+			}
+
+			public Func<bool> AddCondition(Func<bool> condition) {
+				if (this.condition == null)
+					return this.condition = condition;
+				return this.condition = (Func<bool>)Delegate.Combine(this.condition, condition);
+			}
+
+			public bool GetCondition() {
+				return condition?.Invoke() ?? true;
+			}
+		}
+
+		private static readonly Item emptyInstance = new();
+
+		private static object cache;
+
+		private readonly List<Group> items;
+		private readonly List<int> candidatesByType;
+		private readonly List<Predicate<Item>> candidatesByMatch;
+
+		public Group this[int item] { // TODO: ensure that you can modify item instance and add new conditions to it.
+			get {
+				cache = item;
+				return items.Find(x => x.item.type.Equals(item));
+			}
+		}
+
+		public Group this[Index index] {
+			get {
+				return items[index];
+			}
+		}
+
+		public ChestLoot() {
+			cache = null;
+			items = new();
+			candidatesByType = new();
+			candidatesByMatch = new();
+		}
+
+		private static bool CheckInvalidness(int item) {
+			return item < ItemID.None || item >= ItemLoader.ItemCount;
+		}
+
+		private static bool CheckInvalidness(Item item) {
+			ArgumentNullException.ThrowIfNull(item, nameof(item));
+			return CheckInvalidness(item.type);
+		}
+
+		public bool Has(int item) {
+			if (item == 0)
+				return false;
+
+			cache = item;
+			if (CheckInvalidness(item))
+				return false;
+			return items.FindIndex(x => x.item.Equals(cache)) != -1;
+		}
+
+		public bool Has(Item item) {
+			return Has(item.type);
+		}
+
+		public bool Has(Predicate<Item> match) {
+			cache = match;
+			return items.FindIndex(x => ((Predicate<Item>)cache)(x.item)) != -1;
+		}
+
+		public bool Put(int item, Func<bool> condition = null) {
+			return Put(ContentSamples.ItemsByType[item], condition);
+		}
+
+		public bool PutAt(int index, int item, Func<bool> condition = null) {
+			return PutAt(index, ContentSamples.ItemsByType[item], condition);
+		}
+
+		public bool PutAt(Predicate<Item> match, int item, Func<bool> condition = null) {
+			return PutAt(match, ContentSamples.ItemsByType[item], condition);
+		}
+
+		public bool Put(Item item, Func<bool> condition = null) {
+			if (Has(item)) {
+				return false;
+			}
+
+			items.Add(new(item, condition));
+			return true;
+		}
+
+		public bool PutAt(int index, Item item, Func<bool> condition = null) {
```

Comment:
Modders should not have to deal with indexes. Especially since they change with other insertions and removals. I suggest removing all the `index` based overloads.